independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/AnnotationsTransformationContext.java

```diff
@@ -57,12 +67,29 @@ void setAnnotations(C annotations) {
     }
 
     public Collection<AnnotationInstance> getAllAnnotations() {
+        return getAllAnnotationForTarget(getTarget());
+    }
+
+    public Collection<AnnotationInstance> getAllAnnotationTargetAnnotations() {
+        return getAllAnnotationForTarget(getAnnotationTarget());
+    }
+
+    private Collection<AnnotationInstance> getAllAnnotationForTarget(AnnotationTarget target) {
         AnnotationStore annotationStore = get(BuildExtension.Key.ANNOTATION_STORE);
         if (annotationStore == null) {
             throw new IllegalStateException(
                     "Attempted to use the getAllAnnotations() method but AnnotationStore wasn't initialized.");
         }
-        return annotationStore.getAnnotations(getTarget());
+        // Jandex overlay in compatible mode won't allow us to query for METHOD_PARAMETER
+        // hence if the "target" is method param, we query whole method and filter it
+        if (AnnotationTarget.Kind.METHOD_PARAMETER.equals(target.kind())) {
+            return annotationStore.getAnnotations(target.asMethodParameter().method()).stream()
+                    .filter(ai -> AnnotationTarget.Kind.METHOD_PARAMETER.equals(ai.target().kind())
+                            && ai.target().equals(target))
```

Comment:
I'm not sure this is gonna work, but comparing the `position()` should be enough.

Reply:
Why wouldn't this work?
I get the idea with position and can do that instead, I just don't see the issue with equals since annotation targets implement it.

Reply:
So `MethodParameterInfo` and `MethodInfo` do implement `equals()`, but `ClassInfo` does not. It probably doesn't matter in this case, but I'm generally wary about using `AnnotationTarget.equals()`.

Reply:
Yes, `ClassInfo` doesn't matter here but good point, I didn't realize not all ATs have it.

Reply:
`ClassInfo.equals()` is used by `MethodInfo.equals()` (which is used by `MethodParameterInfo.equals()`), so it technically does matter, but it probably doesn't make a big difference here.